### DISCUSSION TITLE
Add function to find nearest color given an array of colors based on color difference using specified color spaces

### DISCUSSION THREAD

**Discussion by lulunac27a** [color-js/color.js#512] [Ideas]
I want to add a function that finds the nearest color from an array of colors that are based on color difference using specified color spaces.

**Comment by facelessuser** [comment]
It's pretty easy to write one around the existing API, but I think it would get a lot of use if it was made part of the existing API, so I'm a +1 on this.

**Comment by lulunac27a** [comment]
The function is defined as:
```js
Color.nearestColor(color, colorArray, colorSpace, weights)
```
`color` is the color to get the closest to that color in the array
`colorArray` is the list of colors to find the closest color
`colorSpace` is the color space method to get the closest color
`weights` are the weights for the color difference based on the color space selected

Example code:
```js
const red = new Color("red");
const blue = new Color("blue");
const green = new Color("green");
const black = new Color("black");
Color.nearestColor("brown", [red, blue, green, black], "lab", [1, 1, 1]);
```

**Comment by LeaVerou** [comment]
Why use a color space for color difference (especially given many are not perceptually uniform) and not a deltaE method which was designed for this?

I’m not sure we need a method for this, but I would support having this a "recipe" in the docs that people could copy paste, or a plugin. It's basically:

```js
function (color, colorArray, {deltaEMethod} = {}) {
	let minDeltaE = Infinity, ret;
	for (let c of colorArray) {
		let deltaE = deltaE(color, c, deltaEMethod);
		if (deltaE < minDeltaE) {
			minDeltaE = deltaE;
			ret = c;
		}
	}
	return ret;
}
```

**Reply by facelessuser** [reply]
I agree, I would be against using a space instead of requiring a specific delta E method. If you had a non-perceptual differencing approach that, for whatever reason, accepted a space and an option, I guess you could pass options to that method, but I wouldn't make specifying the space as part of the API for `nearestColor` or whatever it gets called.

**Reply by lulunac27a** [reply]
The code above is based on the Delta E methods, with the latest one is Delta E 2000.
